### PR TITLE
simulator: Fix flux_request_decode() usage

### PR DESCRIPTION
**Pull request by chu11**
Adjust flux_request_decode() usage given behavior change in flux-core.
In particular, check json_str for non-NULL after calling to ensure
payload was received.

### CODE REVIEW

**Comment by chu11** [comment]
This PR is dependent on the changes in Flux core https://github.com/flux-framework/flux-core/pull/1009

**Comment by garlick** [comment]
flux-framework/flux-core#1009 has been merged so this should probably go in.

**Comment by lipari** [comment]
Yes, I was looking to do this, but the Travis build failed.

**Comment by grondo** [comment]
Eww, yeah there were segfaults in `flux start` under travis on this PR. Let me restart the build and see if it goes better this time.

**Comment by chu11** [comment]
@grondo I had restarted it after Jim pulled in https://github.com/flux-framework/flux-core/pull/1009.  The errors weren't there before, so presumably the merge caused something.  Looking into it now.

I do get errors running 'make check' locally on opal, which were not there before.  Argh ...

**Comment by grondo** [comment]
Yeah, unfortunately corefiles were broken again on Travis so all we're getting is this:
If you can reproduce it locally that's good
```
*** glibc detected *** /home/travis/local2/libexec/flux/cmd/flux-broker: corrupted double-linked list: 0x00002ab412b743d0 ***

======= Backtrace: =========

/lib/x86_64-linux-gnu/libc.so.6(+0x7da66)[0x2ab3cf988a66]

/lib/x86_64-linux-gnu/libc.so.6(+0x7dd98)[0x2ab3cf988d98]

/lib/x86_64-linux-gnu/libc.so.6(+0x7ea69)[0x2ab3cf989a69]

/home/travis/local/lib/libzmq.so.5(+0x2f251)[0x2ab3cf0a9251]

/home/travis/local/lib/libzmq.so.5(+0x2d212)[0x2ab3cf0a7212]

/home/travis/local/lib/libzmq.so.5(+0x39a1c)[0x2ab3cf0b3a1c]

/home/travis/local/lib/libzmq.so.5(+0x39f0d)[0x2ab3cf0b3f0d]

/home/travis/local/lib/libzmq.so.5(+0x1ef1e)[0x2ab3cf098f1e]

/home/travis/local/lib/libzmq.so.5(+0x4b8c6)[0x2ab3cf0c58c6]

/lib/x86_64-linux-gnu/libpthread.so.0(+0x7e9a)[0x2ab3cf2e9e9a]

/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x2ab3cfa010cd]

======= Memory map: ========
```

**Comment by chu11** [comment]
Backtrace on opal

```
(gdb) bt
#0  0x00002aaaab5c9353 in zhash_destroy () from /lib64/libczmq.so.3
#1  0x00002aaac3d03759 in free_simstate (sim_state=0x2aaaf2276c90) at simulator.c:54
#2  0x00002aaac3cf9806 in freectx (arg=0x2aaaf000cad0) at sched.c:317
#3  0x00002aaaab5c92c2 in s_item_destroy () from /lib64/libczmq.so.3
#4  0x00002aaaab5c9364 in zhash_destroy () from /lib64/libczmq.so.3
#5  0x00002aaaaaf209b9 in flux_handle_destroy (hp=hp@entry=0x2aaaec1ffd28) at handle.c:401
#6  0x00002aaaaaf20a2f in flux_close (h=0x2aaaf0000a90) at handle.c:366
#7  0x000000000040a9b3 in module_thread (arg=0x65b110) at module.c:194
#8  0x00002aaaabaa6dc5 in start_thread (arg=0x2aaaec200700) at pthread_create.c:308
#9  0x00002aaaac4bb73d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:113
```

**Comment by grondo** [comment]
Oh, is this the first test against flux-sched since the improved shutdown code went in? If you have time it might be nice to do a git bisect on this, but I'm wondering if there's some missing shutdown path in the simulator now...

**Comment by garlick** [comment]
Uh oh that seems likely.  I'll have a look tomorrow if you don't beat me to it.

**Comment by dongahn** [comment]
Is the problem deterministic? Could it just be not assigning NULL to the sim_state (e.g., at [here](https://github.com/flux-framework/flux-sched/blob/cd04bbd7d6be2e3a3de78c2545df0d765ba11310/sched/sched.c#L935))?

**Comment by chu11** [comment]
@grondo Yeah, I definitely did my first tests/pushes of the RPC patches before the shutdown merge went in.

**Comment by lipari** [comment]
> Is the problem deterministic? Could it just be not assigning NULL to the sim_state (e.g., at here)?

That _does_ appear to fix it:
```
diff --git a/sched/sched.c b/sched/sched.c
index 25e5cc1..f2e6e3e 100644
--- a/sched/sched.c
+++ b/sched/sched.c
@@ -932,6 +932,7 @@ static void trigger_cb (flux_t *h,
     send_reply_request (h, "sched", ctx->sctx.sim_state);
 
     free_simstate (ctx->sctx.sim_state);
+    ctx->sctx.sim_state = NULL;
     Jput (o);
 }
```
with this additional change for good measure:
```
diff --git a/simulator/sim_execsrv.c b/simulator/sim_execsrv.c
index 0bd191a..d18718e 100644
--- a/simulator/sim_execsrv.c
+++ b/simulator/sim_execsrv.c
@@ -602,6 +602,7 @@ static void trigger_cb (flux_t *h,
 
     // Cleanup
     free_simstate (ctx->sim_state);
+    ctx->sim_state = NULL;
     Jput (o);
     zhash_destroy (&job_hash);
 }
```

**Comment by chu11** [comment]
some git bisecting said

```
commit d5378ca47934b0a1bf3838acf2df6245d5318478
Author: Jim Garlick <garlick.jim@gmail.com>
Date:   Tue Mar 14 16:49:30 2017 -0700

    broker: stop the reactor after shutdown timeout
    
    When the shutdown timeout expires, instead of exiting
    directly, tell the reactor to stop so main() can
    exit through its stack of destructors.
```

was the culprit.  Don't know the code 100% and its history, but if broker used to just ```exit()```, but is now calling a shutdown path for each module, perhaps the above shutdown path has always had this corner case bug?

**Comment by garlick** [comment]
Right - that commit is part of flux-framework/flux-core#1005 and yes indeed, modules left unloaded after the test script exits and rc3 runs are now unloaded in the cleanup code.

Ideally, test scripts should have a `flux module remove` for every `flux module load`, since leftover modules will be unloaded without the reactor running, under a timeout, which may not be appropriate for all modules. (If there's going to be a broker segfault i don't know if there is any advantage to having it occur while the test script is running or not.)

I opened #239 to track the fact that some sched test scripts don't unload the modules they load.  Here we should probably try to focus on the segfault.  If @lipari's change fixes it, perhaps @chu11 could tack that on to his PR to get it past travis?

**Comment by lipari** [comment]
@chu11, if prefer, I can create a PR with just those two additional lines and we can merge that before this PR.

**Comment by chu11** [comment]
@lipari How about a separate PR since it's independent of this one.  Thanks!

**Comment by lipari** [comment]
> @lipari How about a separate PR since it's independent of this one. Thanks!

https://github.com/flux-framework/flux-sched/pull/240 ready for review/merging.

**Comment by chu11** [comment]
Code coverage I think is ok.  If I'm reading it correctly, all it's saying is if the new ```!json_str``` checks were never true.  Which is probably to be expected.

**Comment by lipari** [comment]
Thank you, @chu11!
